### PR TITLE
Algorithm

### PR DESCRIPTION
**Pull request by Jeronics**
I created Question and Quiz objects. And finished a first algorithm.

### CODE REVIEW

File: PaintingTrivia.py [hunk 1]
```diff
@@ -116,6 +120,21 @@ def main_menu():
...
 
+@app.route('/results', methods=['GET', 'POST'])
+def show_quiz_results():
+    if request.method == 'POST':
+        # Clear session
+        session.pop('quiz')
+        x = layout_buttons()
+        if x:
+            return x
+    quiz = pickle.loads(session['quiz'])
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: i don't like it. i think we need to install yet another library http://flask.pocoo.org/docs/0.10/api/#sessions

File: PaintingTrivia.py [hunk 2]
```diff
@@ -116,6 +120,21 @@ def main_menu():
...
+    return render_template('show_quiz_results.html',
+                           right_guesses=sum(quiz.score),
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: why not pass the quiz object directly to the template ??

File: PaintingTrivia.py [hunk 3]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
             return x
 
         try:
-            chosen_painter = int(request.form['chosen_painter'])
+            quiz = pickle.loads(session['quiz'])
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: same here. the problem is that, as it turns out, the default session mechanism in flask uses cookies to store information. that means it's stored on the client-side, what we need is a server-side storage...maybe later

File: PaintingTrivia.py [hunk 4]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
+            quiz.process_answer(question, chosen_painter_id == question.correct_option.id)
+            session['quiz'] = pickle.dumps(quiz)
         except KeyError:
-            chosen_painter = -1
-        key = 'right_guesses' if chosen_painter == session[
-            'selected_painter_id'] else 'wrong_guesses'
-        session[key] += 1
+            print "Something failed"
 
-    #http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: why did you delete these comments? I mean, i've put them in here on purpose as a reference (for developers only of course). it does not belong to documentation

File: PaintingTrivia.py [hunk 5]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
-    #the instructions below correspond to the following sql statement:
-    # SELECT painter.id AS painter_id, painter.name AS painter_name FROM painter
-    # WHERE EXISTS (SELECT * FROM painting WHERE painter.id = painting.painter_id)
-    # ORDER BY random() LIMIT 4
+    if session['quiz'] is None:
+        # Obtain the painters and the paintings randomly
+        selected_painters, selected_paintings = initialize_images(num_painters=10)
+        quiz = PainterQuiz(selected_paintings, selected_painters)
+        session['quiz'] = pickle.dumps(quiz)
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: again, i don't like pickle.dumps/loads calls. it should be abstracted away....

File: PaintingTrivia.py [hunk 6]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
+    session['quiz'] = pickle.dumps(quiz)
+    if not question:
+        return redirect(url_for('show_quiz_results'))
     return render_template('guess_the_painter.html',
-                           painters=painters,
-                           selected_painter=selected_painter,
-                           selected_painting=selected_painting,
-                           right_guesses=session['right_guesses'],
-                           wrong_guesses=session['wrong_guesses'],
+                           painters=question.option_list,
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: just pass the whole question object to the template

File: PaintingTrivia.py [hunk 7]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
 
+def initialize_images(num_painters=10):
+    stmt = exists().where(Painter.id == Painting.painter_id)
+    selected_painters = Painter.query.filter(stmt).order_by(func.random()).limit(num_painters).all()
+
+    painters_dict = dict()
+    paintings_dict = dict()
+    selected_paintings = []
+    for key, painter in enumerate(selected_painters):
+        paintings_aux = Painting.query.filter(Painting.painter_id == painter.id).order_by(func.random()).limit(1).all()
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: one() instead of all()

File: PaintingTrivia.py [hunk 8]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
+        selected_paintings.extend(paintings_aux)
+        painters_dict[key] = painter
+        paintings_dict[key] = paintings_aux[0]  # get the first
+    return painters_dict, paintings_dict
+
+
+def initialize_saint_images(num_saints=10):
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: This function is almost one-to-one copy of the previous one. please refactor to have only one

File: PaintingTrivia.py [hunk 9]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
+
+    painters_dict = dict()
+    paintings_dict = dict()
+    selected_paintings = []
+    for key, painter in enumerate(selected_painters):
+        paintings_aux = Painting.query.filter(Painting.painter_id == painter.id).order_by(func.random()).limit(1).all()
+        selected_paintings.extend(paintings_aux)
+        painters_dict[key] = painter
+        paintings_dict[key] = paintings_aux[0]  # get the first
+    return painters_dict, paintings_dict
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: i don't actually understand the sense of this method. you are creating 2 dictionaries (sort of a mapping i guess), but what for? 
```
painter_dict{
 1 : painter1
}
painting_dict{
 1 : painting_of_painter1
}
```

File: PaintingTrivia.py [hunk 10]
```diff
@@ -161,38 +211,42 @@ def guess_the_saint():
         x = layout_buttons()
         if x:
             return x
+
         try:
-            chosen_painting = int(request.form['chosen_saint'])
+            quiz = pickle.loads(session['quiz'])
+            chosen_saint_id = int(request.form['chosen_saint'])
+            question = quiz.current_question
+            quiz.process_answer(question, chosen_saint_id == question.correct_option.id)
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: the quiz should know (and store) what correct answer is. so you shouldn't pass it to the object from the outside

File: Quiz.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
+import random
```
Comment (denisergashbaev) [inline Quiz.py]: in general, good idea with objects !

File: Quiz.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+from abc import ABCMeta, abstractmethod
+
+
```
Comment (denisergashbaev) [inline Quiz.py]: the file name Quiz.py should be lowercase

File: Quiz.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+class MultipleChoiceQuestion:
+    def __init__(self, element, correct_option, option_list):
+        self.element = element
+        self.correct_option = correct_option
+        self.option_list = option_list
+
+
+class Quiz:
+    __metaclass__ = ABCMeta
```
Comment (denisergashbaev) [inline Quiz.py]: cool!

File: PaintingTrivia.py (hunk 3 above)
Comment (Jeronics) [inline PaintingTrivia.py]: Ohhh i get it.. I thought there was only one storage. Okk ill check it out!

File: PaintingTrivia.py (hunk 2 above)
Comment (Jeronics) [inline PaintingTrivia.py]: Okk! on it!

File: PaintingTrivia.py [hunk 14]
```diff
@@ -125,35 +144,66 @@ def guess_the_painter():
...
 
-    #http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
-    #the instructions below correspond to the following sql statement:
-    # SELECT painter.id AS painter_id, painter.name AS painter_name FROM painter
-    # WHERE EXISTS (SELECT * FROM painting WHERE painter.id = painting.painter_id)
-    # ORDER BY random() LIMIT 4
+    if session['quiz'] is None:
+        # Obtain the painters and the paintings randomly
+        selected_painters, selected_paintings = initialize_images(num_painters=10)
+        quiz = PainterQuiz(selected_paintings, selected_painters)
```
Comment (Jeronics) [inline PaintingTrivia.py]: Done

File: PaintingTrivia.py (hunk 7 above)
Comment (Jeronics) [inline PaintingTrivia.py]: done

File: PaintingTrivia.py (hunk 9 above)
Comment (Jeronics) [inline PaintingTrivia.py]: Its for quickly checking if its correct. I'll tell you in a hangout.

File: PaintingTrivia.py (hunk 10 above)
Comment (Jeronics) [inline PaintingTrivia.py]: True! What was I thinking!?

File: Quiz.py (hunk 12 above)
Comment (Jeronics) [inline Quiz.py]: Done

File: Quiz.py (hunk 11 above)
Comment (Jeronics) [inline Quiz.py]: Thanks!

File: Quiz.py (hunk 13 above)
Comment (Jeronics) [inline Quiz.py]: ;)

File: Quiz.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+    def __init__(self):
+        self.number_of_rounds = 0
+        self.score = []
+
+        # current
+        self.current_question = None
+
+    def getScore(self):
```
Comment (Jeronics) [inline Quiz.py]: done!

File: Quiz.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+    def next_question(self):
+        self.number_of_rounds += 1
+        question = self.generate_question()
+        return question
+
+
+class ImageQuiz(Quiz):
+    def __init__(self, elements_dict, options_dict, num_elements_quiz=4):
+        super(ImageQuiz, self).__init__()
```
Comment (Jeronics) [inline Quiz.py]: gracias! ;)

File: Quiz.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        quiz_options_list.extend(random.sample(other_options, self.num_elements_quiz - 1))
+
+        # 3. Shuffle the new sub_option_list
+        random.shuffle(quiz_options_list)
+        question = MultipleChoiceQuestion({correct_key: correct_element}, correct_option, quiz_options_list)
+
+        self.current_question = question
+        return question
+
+    def process_answer(self, question, correct):
```
Comment (Jeronics) [inline Quiz.py]: Do you mean the process_answer method??

File: logic/test_algorithm.py [hunk 18]
```diff
@@ -0,0 +1,4 @@
+
+
+class Algorithm:
```
Comment (Jeronics) [inline logic/test_algorithm.py]: Yup!

**Comment by Jeronics** [comment]
I fixed the comments! We have to figure out the server side variable still

**Comment by Jeronics** [comment]
Ok!! I fixed the bug!! The First step of logic is done! Merge that sh***t!

File: PaintingTrivia.py (hunk 3 above)
Comment (denisergashbaev) [inline PaintingTrivia.py]: afaik, following is the _pythonic_ way to get the values out of the dictionary. please change
`    try:
        quiz = pickle.loads(session['quiz'])
    except KeyError:
        quiz = PainterQuiz()
        session['quiz'] = pickle.dumps(quiz)
`

File: PaintingTrivia.py [hunk 19]
```diff
@@ -125,32 +139,34 @@ def guess_the_painter():
...
         except KeyError:
-            chosen_painter = -1
-        key = 'right_guesses' if chosen_painter == session[
-            'selected_painter_id'] else 'wrong_guesses'
-        session[key] += 1
-
-    #http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
-    #the instructions below correspond to the following sql statement:
-    # SELECT painter.id AS painter_id, painter.name AS painter_name FROM painter
+            print "Something failed"
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: bad. by suppressing the error with a meaningless message "Something failed" you neither gain important information (what exactly failed) nor do you recover from the error correctly. You need to put into try/except block only the stuff that can break (in that case it is getting the chosen_painter from the html) and setting it to a default value "-1" could be ok. please change here and elsewhere

File: templates/guess_the_painter.html [hunk 20]
```diff
@@ -1,37 +1,37 @@
...
-        <div>Wrong guesses: {{ wrong_guesses }}</div>
-        <div>Who is this painting created by?</div>
+{% include "login_register_segment.html"%}
+<div class="page">
+    <h1>PaintingTrivia</h1>
+    {% for message in get_flashed_messages() %}
+    <div class="flash">{{ message }}</div>
+    {% endfor %}
+    <div>Statistics:</div>
+    <div>Right guesses: {{ quiz.score|sum() }}</div>
```
Comment (denisergashbaev) [inline templates/guess_the_painter.html]: just create a method in quiz that is gonna provide you information about the right and wrong guesses. no need to put logic into templates

File: templates/guess_the_painter.html [hunk 21]
```diff
@@ -1,37 +1,37 @@
...
+    <div>Wrong guesses: {{ quiz.score|length - quiz.score|sum() }}</div>
+    <div>Who is this painting created by?</div>
 
-        <h2>{{ selected_painting.title }}</h2>
+    <h2>{{ quiz.current_question.element.values()[0].title }}</h2>
```
Comment (denisergashbaev) [inline templates/guess_the_painter.html]: `quiz.current_question.element.values()[0].title` -- this is very ugly. when i look at it i have no idea what element, values(), [0] mean. please come up with a prettier solution

File: templates/guess_the_saint.html [hunk 22]
```diff
@@ -1,34 +1,37 @@
...
-        <div>Wrong guesses: {{ wrong_guesses }}</div>
-        <div>Who is this painting created by?</div>
+{% include "login_register_segment.html"%}
+<div class="page">
+    <h1>PaintingTrivia</h1>
+    {% for message in get_flashed_messages() %}
+    <div class="flash">{{ message }}</div>
+    {% endfor %}
+    <div>Statistics:</div>
+    <div>Right guesses: {{ quiz.score|sum() }}</div>
```
Comment (denisergashbaev) [inline templates/guess_the_saint.html]: same

File: templates/guess_the_saint.html [hunk 23]
```diff
@@ -1,34 +1,37 @@
...
-                <button id="chosen_saint_{{ saint.id }}" type="button" name="chosen_saint"
-                        value="{{ saint.id }}"  onclick="app.submitGuess(this, this.form, {{ selected_saint.id }})">
-                    {{ saint.name }}
-                </button>
-                <br/>
-                {% endfor %}
-            </ul>
-        </form>
+    <div class="painting"><img class="painting"
+                               src="{{ url_for('static', filename = 'pics/%s' % quiz.current_question.element.values()[0].file_name) }}"
```
Comment (denisergashbaev) [inline templates/guess_the_saint.html]: same

File: PaintingTrivia.py [hunk 24]
```diff
@@ -161,38 +177,36 @@ def guess_the_saint():
...
-                           selected_painting=selected_painting,
-                           right_guesses=session['right_guesses'],
-                           wrong_guesses=session['wrong_guesses'],
+                           quiz=quiz,
+                           selected_painter=Painter.query.filter(Painter.id == chosen_painting.painter_id).first(),
                            username=session['username'])
 
 
 if __name__ == '__main__':
-    # run on the machine ip address (local network)
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: please restore these comments

File: PaintingTrivia.py [hunk 25]
```diff
@@ -125,32 +139,34 @@ def guess_the_painter():
...
-        session[key] += 1
-
-    #http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
-    #the instructions below correspond to the following sql statement:
-    # SELECT painter.id AS painter_id, painter.name AS painter_name FROM painter
+            print "Something failed"
+
+            # http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#using-exists
+
+    # the instructions below correspond to the following sql statement:
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: of course, these comments do not belong in here anymore  ;)

File: PaintingTrivia.py [hunk 26]
```diff
@@ -161,38 +177,36 @@ def guess_the_saint():
...
+    question = quiz.next_question()
 
-    saints_list = Saint.query.filter(Saint.paintings.any()).limit(4).all()
-    selected_saint = random.choice(saints_list)
+    session['quiz'] = pickle.dumps(quiz)
+    if not question:
+        return redirect(url_for('show_quiz_results'))
 
-    selected_painting = Painting.query.filter(Painting.saints.any(id=selected_saint.id)).first()
+    chosen_painting = question.element.values()[0]
```
Comment (denisergashbaev) [inline PaintingTrivia.py]: this is also very ugly

**Comment by denisergashbaev** [comment]
@Jeronics, nope. see my last comments, it's not good enough yet

File: templates/guess_the_painter.html (hunk 20 above)
Comment (Jeronics) [inline templates/guess_the_painter.html]: you're so smart denis!! :+1:

File: templates/guess_the_painter.html (hunk 21 above)
Comment (Jeronics) [inline templates/guess_the_painter.html]: Will do captain!

File: quiz.py [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+
+
+class ImageQuiz(Quiz):
+    def __init__(self, num_elements_quiz=4):
+        super(ImageQuiz, self).__init__()
+
+        elements_dict, options_dict = self.initialize_image_quiz(num_elements=10)
+        self.elements_dict = elements_dict
+        self.options_dict = options_dict
+        self.num_elements_quiz = num_elements_quiz
```
Comment (denisergashbaev) [inline quiz.py]: why not ?
`self.elements_dict, self.options_dict = self.initialize_image_quiz(num_elements=10)`

File: quiz.py [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+from sqlalchemy.sql.expression import exists
+from sqlalchemy.sql.functions import func
+from models.painter import Painter
+from models.painting import Painting
+from models.saint import Saint
+
+
+class MultipleChoiceQuestion:
+    def __init__(self, element, correct_option, option_list):
+        self._element = element
```
Comment (denisergashbaev) [inline quiz.py]: what does element or _element mean? please give the variables a proper self-explanatory name

File: quiz.py [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+        self.number_of_rounds = 0
+        self.score = []
+
+        # current
+        self.current_question = None
+
+        self.right_guesses = sum(self.score)
+        self.wrong_guesses = len(self.score) - self.right_guesses
+
+    def get_score(self):
```
Comment (denisergashbaev) [inline quiz.py]: this getter is not used at all in code. please remove. also, afaik, it's not pythonic to create getters, see: 
http://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters
http://stackoverflow.com/questions/6618002/python-property-versus-getters-and-setters

File: quiz.py [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+        return self.score
+
+    @abstractmethod
+    def generate_question(self):
+        return None
```
Comment (denisergashbaev) [inline quiz.py]: just return is enough 
see https://pymotw.com/2/abc/

File: quiz.py [hunk 31]
```diff
@@ -0,0 +1,148 @@
...
+    def update_score(self, correct):
+        if correct:
+            self.right_guesses += 1
+        else:
+            self.wrong_guesses += 1
+        self.score.append(correct)
+
+    def next_question(self):
+        self.number_of_rounds += 1
+        question = self.generate_question()
```
Comment (denisergashbaev) [inline quiz.py]: good

File: quiz.py [hunk 32]
```diff
@@ -0,0 +1,148 @@
...
+        elements_dict, options_dict = self.initialize_image_quiz(num_elements=10)
+        self.elements_dict = elements_dict
+        self.options_dict = options_dict
+        self.num_elements_quiz = num_elements_quiz
+
+        # Auxiliary attributes
+        self.seen_elements_dict = dict()
+        self.current_question = None
+
+    def initialize_image_quiz(self, num_elements):
```
Comment (denisergashbaev) [inline quiz.py]: why it is not marked as abstract method (with empty return), since you cant provide a default implementation here?

File: quiz.py [hunk 33]
```diff
@@ -0,0 +1,148 @@
...
+                self.elements_dict = self.seen_elements_dict
+                self.seen_elements_dict = dict()
+
+        # 1. grab a key from the element_list randomly
+        correct_key = random.choice(self.elements_dict.keys())
+
+        # 2. Create a subset of the option_list for the quiz
+        correct_element = self.elements_dict[correct_key]
+        correct_option = self.options_dict[correct_key]
+        quiz_options_list = list()
```
Comment (denisergashbaev) [inline quiz.py]: [] instead of list()

File: quiz.py [hunk 34]
```diff
@@ -0,0 +1,148 @@
...
+        quiz_options_list.append(correct_option)
```
Comment (denisergashbaev) [inline quiz.py]: well, you can just initialize and append simultaneously =[correct_option]

File: quiz.py [hunk 35]
```diff
@@ -0,0 +1,148 @@
...
+        other_options = filter(lambda x: x != correct_option, set(self.options_dict.values()))
```
Comment (denisergashbaev) [inline quiz.py]: good re filter()

File: PaintingTrivia.py (hunk 24 above)
Comment (denisergashbaev) [inline PaintingTrivia.py]: ?

File: quiz.py [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+class ImageQuiz(Quiz):
+    def __init__(self, num_elements_quiz=4):
+        super(ImageQuiz, self).__init__()
+
+        elements_dict, options_dict = self.initialize_image_quiz(num_elements=10)
+        self.elements_dict = elements_dict
+        self.options_dict = options_dict
+        self.num_elements_quiz = num_elements_quiz
+
+        # Auxiliary attributes
```
Comment (denisergashbaev) [inline quiz.py]: well, it's not a very telling comment...

**Comment by denisergashbaev** [comment]
@Jeronics so, still stuff to do. see my comments. we want a clean code here ;)

File: quiz.py (hunk 27 above)
Comment (Jeronics) [inline quiz.py]: true! :sake:

File: quiz.py (hunk 28 above)
Comment (Jeronics) [inline quiz.py]: 1) I'll look into it!
2) I DO want _element to be private
3) Its complicated to explain. Its easier to keep a key of the correct value. It worked best when I layed out the possibilities!

Comment (denisergashbaev) [inline quiz.py]: Call them like 'question' and 'answer'.
If u make smth private, then you should only access it inside of the class, not feom anywhere outside

File: quiz.py (hunk 30 above)
Comment (denisergashbaev) [inline quiz.py]: Also, you, don't really update right/wrong_guesses. This is dynamic info, make a property out of it (@property) instead, no need for calculation in the constructor

File: quiz.py (hunk 28 above)
Comment (Jeronics) [inline quiz.py]: 1.) I'll give proper names
2.) It is supposed to be private. Where is it outside?
3.) Long story short, its better for checking with index.

File: quiz.py (hunk 29 above)
Comment (Jeronics) [inline quiz.py]: gone! :dancer:

File: quiz.py (hunk 30 above)
Comment (Jeronics) [inline quiz.py]: Yes sir :racehorse:

File: quiz.py (hunk 31 above)
Comment (Jeronics) [inline quiz.py]: thanks :beers:

File: quiz.py (hunk 32 above)
Comment (Jeronics) [inline quiz.py]: Ok!!

File: quiz.py (hunk 33 above)
Comment (Jeronics) [inline quiz.py]: Really? why?

File: quiz.py (hunk 34 above)
Comment (Jeronics) [inline quiz.py]: True!

**Comment by denisergashbaev** [comment]
@Jeronics is it all done now? Can i review?

File: quiz.py [hunk 37]
```diff
@@ -0,0 +1,144 @@
...
+    def initialize_image_quiz(self, num_elements=5):
+        selected_saints = Saint.query.filter(Saint.paintings.any()).order_by(func.random()).limit(num_elements).all()
+
+        saints_dict = dict()
+        paintings_dict = dict()
+        selected_paintings = []
+        for key, saint in enumerate(selected_saints):
+            paintings_aux = Painting.query.filter(Painting.saints.any(id=saint.id)).order_by(func.random()).limit(
+                1).first()
+            selected_paintings.extend([paintings_aux])
```
Comment (denisergashbaev) [inline quiz.py]: why not `selected_paintings.add(paintings_aux)` since you are only adding one element??

File: quiz.py [hunk 38]
```diff
@@ -0,0 +1,144 @@
...
+        # 1. grab a key from the element_list randomly
+        correct_key = random.choice(self.elements_dict.keys())
+
+        # 2. Create a subset of the option_list for the quiz
+        correct_element = self.elements_dict[correct_key]
+        correct_option = self.options_dict[correct_key]
+        other_options = filter(lambda x: x != correct_option, set(self.options_dict.values()))
+        aux_list = random.sample(other_options, self.num_elements_quiz - 1)
+        quiz_options_set = set(aux_list)
+        quiz_options_set.add(correct_option)
```
Comment (denisergashbaev) [inline quiz.py]: ok, i think i was wrong last time ;)

File: quiz.py [hunk 39]
```diff
@@ -0,0 +1,144 @@
...
+
+    def __init__(self):
+        self.number_of_rounds = 0
+        self.score = []
+
+        # current
+        self.current_question = None
+
+        self.right_guesses = sum(self.score)
+        self.wrong_guesses = len(self.score) - self.right_guesses
```
Comment (denisergashbaev) [inline quiz.py]: this initialization of `right/wrong_guesses` is actually useless, because the `score` is always empty in the beginning

File: quiz.py [hunk 40]
```diff
@@ -0,0 +1,144 @@
...
+from models.painting import Painting
+from models.saint import Saint
+
+
+class MultipleChoiceQuestion:
+    def __init__(self, question, correct_answer, option_list):
+        self.question_dict = question
+        self.question = question.values()[0]
+        self.correct_answer = correct_answer
+        self.option_list = option_list
```
Comment (denisergashbaev) [inline quiz.py]: `other_anwers` instead of `option_list` would be more discriptive

File: quiz.py [hunk 41]
```diff
@@ -0,0 +1,144 @@
...
+
+        # update score
+        self.update_score(correct)
+
+        # 0. pop element from element list
+        self.elements_dict.pop(self.current_question.question_dict.keys()[0])
+
+        if not correct:
+            # save popped element in the seen_elements list
+            self.seen_elements_dict.update(self.current_question.question_dict)
```
Comment (denisergashbaev) [inline quiz.py]: cool with update :+1:

File: quiz.py [hunk 42]
```diff
@@ -0,0 +1,144 @@
...
+        stmt = exists().where(Painter.id == Painting.painter_id)
+        selected_painters = Painter.query.filter(stmt).order_by(func.random()).limit(num_elements).all()
+
+        painters_dict = dict()
+        paintings_dict = dict()
+        selected_paintings = []
+        for key, painter in enumerate(selected_painters):
+            paintings_aux = Painting.query.filter(Painting.painter_id == painter.id).order_by(func.random()).limit(
+                1).first()
+            selected_paintings.extend([paintings_aux])
```
Comment (denisergashbaev) [inline quiz.py]: 1) `painting_aux` (singular) instead of `paintings_aux`
2) `append` instead of `extend` since you are only adding one element

File: quiz.py (hunk 37 above)
Comment (Jeronics) [inline quiz.py]: I wanted to make it general enough so that maybe a couple or more paintings can be taken for each saint. Should I change it?

File: quiz.py (hunk 38 above)
Comment (Jeronics) [inline quiz.py]: haha why? Should I put it back?

File: quiz.py (hunk 39 above)
Comment (Jeronics) [inline quiz.py]: okk, I'll put 0 for both on initialization

File: quiz.py (hunk 40 above)
Comment (Jeronics) [inline quiz.py]: other_answers is not accurate either, cause it implies that the correct answer is not included.

File: quiz.py (hunk 42 above)
Comment (Jeronics) [inline quiz.py]: 1) Done
2) Again, I wanted to make it general enough so that maybe a couple or more paintings can be taken for each saint. Should I change it?

**Comment by Jeronics** [comment]
Checkout the comments, I changed everything I agreed with. If you insist on anything I'll change it!

**Comment by denisergashbaev** [comment]
Merged, let's move on
